common/log.h

```diff
@@ -47,6 +50,396 @@
 //
 // --------------------------------
 
+#ifdef LOG_WITH_TEST
+inline void log_test();
+#endif
+
+// Utility for synchronizing log configuration state
+//  since std::optional was introduced only in c++17
+enum class LogTriState
+{
+    Same,
+    False,
+    True
+};
+
+enum class LogTargetChannel
+{
+    Logfile,
+    Screen
+};
+
+class LogStateWrapper
```

Comment:
This class is long enough that it's hard to tell where it ends. I would prefer if the methods were only declared in the class, and defined with `inline` outside of it. Even better, you could put the definitions in a file called log.cpp.

Reply:
> log.cpp

Sounds good, I'll do that.